File: src/main/java/com/evergreenclient/hudmod/utils/Facing.java
/*
 * Copyright (C) Evergreen [2020 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/Evergreen-Client/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package com.evergreenclient.hudmod.utils;

public enum Facing {
    NORTH("North", "N"),
    NORTH_EAST("North East", "NE"),
    EAST("East", "E"),
    SOUTH_EAST("South East", "SE"),
    SOUTH("South", "S"),
    SOUTH_WEST("South West", "SW"),
    WEST("West", "W"),
    NORTH_WEST("North West", "NW");

    private final String big;
    private final String small;
    Facing(String big, String small) {
        this.big = big;
        this.small = small;
    }

    public String getNormal() {
        return this.big;
    }

    public String getAbbreviated() {
        return this.small;
    }
}

File: src/main/kotlin/cc/polyfrost/evergreenhud/utils/Facing.kt
package cc.polyfrost.evergreenhud.utils

import net.minecraft.util.MathHelper

enum class Facing(val full: String, val abbreviated: String) {
    NORTH("North", "N"),
    NORTH_EAST("North East", "NE"),
    EAST("East", "E"),
    SOUTH_EAST("South East", "SE"),
    SOUTH("South", "S"),
    SOUTH_WEST("South West", "SW"),
    WEST("West", "W"),
    NORTH_WEST("North West", "NW");

    companion object {
        fun parse(yaw: Float): Facing {
            val rotationYaw = MathHelper.wrapAngleTo180_float(yaw)

            return when {
                rotationYaw >= 165f || rotationYaw <= -165 -> NORTH
                rotationYaw in -165f..-105f -> NORTH_EAST
                rotationYaw in -105f..-75f -> EAST
                rotationYaw in -75f..-15f -> SOUTH_EAST
                rotationYaw in -15f..15f -> SOUTH
                rotationYaw in 15f..75f -> SOUTH_WEST
                rotationYaw in 75f..105f -> WEST
                rotationYaw in 105f..165f -> NORTH_WEST

                else -> NORTH
            }
        }

        fun parseExact(yaw: Float): Facing {
            val rotationYaw = MathHelper.wrapAngleTo180_float(yaw)

            return when {
                rotationYaw == -180f -> NORTH
                rotationYaw > -180f && rotationYaw < -90f -> NORTH_EAST
                rotationYaw == -90f -> EAST
                rotationYaw > -90f && rotationYaw < 0f -> SOUTH_EAST
                rotationYaw == 0f -> SOUTH
                rotationYaw > 0f && rotationYaw < 90f -> SOUTH_WEST
                rotationYaw == 90f -> WEST
                rotationYaw > 90f && rotationYaw < 180f -> NORTH_WEST

                else -> NORTH
            }
        }
    }
}